HaxmLoader/haxm_loader.c

```diff
@@ -100,55 +99,64 @@ int DriverUninstall(void)
     return retval;
 }
 
-int DriverInstall(char *sysFileName)
+static int DriverInstall(const wchar_t *sysFileName)
 {
     HANDLE fileHandle;
-    UCHAR driverLocation[MAX_PATH];
+    wchar_t driverLocation[MAX_PATH];
     SC_HANDLE hSCManager = NULL;
     SC_HANDLE hService = NULL;
     int retval = 1;
 
-    GetCurrentDirectory(MAX_PATH, (LPTSTR)driverLocation);
-    strcat((char *)driverLocation, "\\");
-    strcat((char *)driverLocation, sysFileName);
+    if (!GetCurrentDirectory(MAX_PATH, (LPTSTR)driverLocation)) {
+        printf("%s(): Failed to get current directory\r\n", __FUNCTION__);
+        goto cleanup;
+    }
+    if (StringCchCatW(driverLocation, MAX_PATH, L"\\") != S_OK ||
+        StringCchCatW(driverLocation, MAX_PATH, sysFileName) != S_OK) {
+        printf("%s(): Failed to get absolute path:\r\n",
+                __FUNCTION__);
+        wprintf(L"  sysFileName='%s'\r\n", sysFileName);
+        goto cleanup;
+    }
 
     // check if sysFile exists
-    fileHandle = CreateFile((LPCSTR)driverLocation,
```

Comment:
I'm not very familiar with unicode coding, my impression is there is a macro "_T" which could handle this as well?

Reply:
`_T` is probably just another form of the `TEXT("some string")` macro, which actually expands to `L"some string"` when `_UNICODE` is defined:

https://docs.microsoft.com/en-us/windows/desktop/learnwin32/working-with-strings

The same MSDN article also deprecates the use of `TEXT`:

> The **TEXT** and **TCHAR** macros are less useful today, because all applications should use Unicode.

That's why I'm using Windows APIs with the `W` suffix explicitly--we don't need to compile the code for non-Unicode versions Windows, because all the Windows versions that HAXM can run on support Unicode. I also tried to follow guidelines from section 10 of this article:

https://utf8everywhere.org/